From c29cc581e959913d401fb3866ac3351b01a39b69 Mon Sep 17 00:00:00 2001
From: Fabrizio Castro <fabrizio.castro.jz@renesas.com>
Date: Wed, 15 Jun 2022 21:07:20 +0100
Subject: [PATCH] rtc: isl1208: Add oscillator selection

By default the ISL1208 will try and use the internal oscillator.
While that's well and good, there are cases where we actually
want to use the external oscillator, but the driver has no
interface for that.

Add the device tree property "external-oscillator" to allow for
external oscillator selection.

Signed-off-by: Fabrizio Castro <fabrizio.castro.jz@renesas.com>
---
 drivers/rtc/rtc-isl1208.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/drivers/rtc/rtc-isl1208.c b/drivers/rtc/rtc-isl1208.c
index ebb691fa48a6..da5e50295539 100644
--- a/drivers/rtc/rtc-isl1208.c
+++ b/drivers/rtc/rtc-isl1208.c
@@ -166,6 +166,26 @@ isl1208_i2c_validate_client(struct i2c_client *client)
 	return 0;
 }
 
+static int
+isl1208_set_external_oscillator(struct i2c_client *client, bool enable)
+{
+	s32 data;
+	int ret;
+
+	data = i2c_smbus_read_byte_data(client, ISL1208_REG_SR);
+	if (data < 0)
+		return data;
+
+	if (enable)
+		ret = i2c_smbus_write_byte_data(client, ISL1208_REG_SR, data |
+						ISL1208_REG_SR_XTOSCB);
+	else
+		ret = i2c_smbus_write_byte_data(client, ISL1208_REG_SR, data &
+						~ISL1208_REG_SR_XTOSCB);
+
+	return ret;
+}
+
 static int
 isl1208_i2c_get_sr(struct i2c_client *client)
 {
@@ -802,6 +822,7 @@ isl1208_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	int rc = 0;
 	struct isl1208_state *isl1208;
 	int evdet_irq = -1;
+	bool external_oscillator;
 
 	if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C))
 		return -ENODEV;
@@ -894,6 +915,18 @@ isl1208_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	if (rc)
 		return rc;
 
+	external_oscillator = of_property_read_bool(client->dev.of_node,
+						    "external-oscillator");
+
+	rc = isl1208_set_external_oscillator(client, external_oscillator);
+	if (rc)
+		return rc;
+
+	if (external_oscillator)
+		dev_info(&client->dev, "Using external oscillator");
+	else
+		dev_info(&client->dev, "Using internal oscillator");
+
 	return rtc_register_device(isl1208->rtc);
 }
 
-- 
2.34.1

